aten/src/ATen/functorch/BatchRulesScatterOps.cpp

```diff
@@ -1056,6 +1056,163 @@ std::tuple<Tensor,optional<int64_t>> masked_fill_scalar_batch_rule(
   return std::make_tuple(result, 0);
 }
 
+std::tuple<Tensor,optional<int64_t>> index_fill_int_scalar_batch_rule_impl(
+    Tensor & self, optional<int64_t> self_bdim,
+    int64_t dim,
+    const Tensor & index, optional<int64_t> index_bdim,
+    const Scalar & value,
+    const bool inplace) {
+    const auto self_logical_rank = rankWithoutBatchDim(self, self_bdim);
+    Tensor self_ = moveBatchDimToFront(self, self_bdim);
+    Tensor index_ = moveBatchDimToFront(index, index_bdim);
+    dim = maybe_wrap_dim(dim, self_logical_rank);
+
+    if (inplace && !self_bdim.has_value()) {
+      vmapIncompatibleInplaceError("index_fill_");
+    }
+
+    if (!index_bdim) {
+      if (self_logical_rank == 0){
+        self_.unsqueeze_(-1);
+      }
+      self_.index_fill_(dim + 1, index_, value);
+      if (self_logical_rank == 0) {
+        self_.squeeze_(-1);
+      }
+      return std::make_tuple(self_, 0);
+    }
+
+    auto batch_size = get_bdim_size2(self, self_bdim, index, index_bdim);
+    self_ = ensure_has_bdim(self_, self_bdim.has_value(), batch_size);
+    index_ = ensure_has_bdim(index_, index_bdim.has_value(), batch_size);
+
+    if (inplace) {
+      // Do for-loop for in-place because we cannot reshape
+      // `self_` having an incompatible stride without copying
+      for (const auto i : c10::irange(0, batch_size)) {
+        const auto& self_slice = self_.select(0, i);
+        const auto& index_slice = index_.select(0, i);
+        self_slice.index_fill_(
+          dim,
+          index_slice,
+          value
+        );
+      }
+      return std::make_tuple(self_, 0);
+    }
+
+    self_ = self_bdim.has_value() ? self_ : self_.clone();
+
+    if (self.dim() > 1){
+      auto index_offset = at::arange(
+        batch_size,
+        at::TensorOptions().dtype(index_.scalar_type()).device(index_.device())
+      );
+      index_ = index_.add(index_offset.unsqueeze(-1), self_.size(dim + 1));
+      index_ = reshape_dim_into(0, 0, index_);
+      self_ = reshape_dim_into(0, dim, self_);
+      self_.index_fill_(dim, index_, value);
+      self_ = reshape_dim_outof(dim, batch_size, self_);
+      return std::make_tuple(self_, dim);
+    }
+
+    // If self.dim() is 0 or 1, the batch dim is certainly 0, and we must apply batched indices to each row.
+    index_ = reshape_dim_into(0, 0, index_);
+    if (self_logical_rank == 0) {
+      self_.unsqueeze_(-1);
+    }
+    self_.index_fill_(dim + 1, index_, value);
+    if (self_logical_rank == 0) {
+      self_.squeeze_(-1);
+    }
+
+    return std::make_tuple(self_, 0);
```

Comment:
I'm not sure the condition here is correct -- it doesn't make sense that we're branching on self.dim() > 1; I would have expected this to look something like:

```py
if (self_logical_rank == 0) {
   ...
   return ...
}

// handle the zero dim case
index_ = reshape_dim_into(0, 0, index_);
self_.unsqueeze_(-1);
self_.index_fill_(dim + 1, index_, value);
self_.squeeze_(-1);
```
Please try out the additional test cases I suggest below.